src/puppetlabs/puppetdb/http/query.clj

```diff
@@ -4,13 +4,44 @@
    Functions that aid in the parsing, serialization, and manipulation
    of PuppetDB queries embedded in HTTP parameters."
   (:require [puppetlabs.puppetdb.cheshire :as json]
-            [clojure.core.match :as cm]))
+            [clojure.core.match :as cm]
+            [puppetlabs.puppetdb.query-eng :refer [produce-streaming-body produce-streaming-body']]
+            [net.cgrand.moustache :refer [app]]
+            [schema.core :as s]
+            [puppetlabs.puppetdb.schema :as pls]
+            [puppetlabs.puppetdb.query.paging :refer [parse-limit' parse-offset' parse-order-by']]
+            [puppetlabs.puppetdb.utils :as utils]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def puppetdb-query-schema
+  "This schema defines a PuppetDB query and it's available
+  parameters. In a GET request this is contained in various query
+  parameters, for POST requests this should be contained in the body
+  of the request"
+  {:query (s/maybe [s/Any])
```

Comment:
given that we support the query=null case, should we go ahead and make the key optional?